sensor/bmi088: Add BMI088 Device Driver.

JerrySlhao opened the pull request:
add bmi088 device used spi,now support spi. By the way, the STM32G4 series SPI driver is added to the SPI driver of STM32.

## Summary

There is no driver of BMI088 in the sensor device driver, I am using this sensor driver, so it is just submitted.

## Impact

This part of the code for the system just adds a new device driver function, convenient for everyone to use.

## Testing

I used STM32G474 to create the board-level driver in the board directory, registered the bmi088 device driver when bringing up, and then created the test program of bmi088 in nuttx-app/example.

### nuttx-apps/examples/bmi088/bmi088_main.c

```
/****************************************************************************
 * apps/system/bmi088/bmi088_main.c
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.  The
 * ASF licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 ****************************************************************************/

/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/config.h>
#include <nuttx/arch.h>

#include <errno.h>
#include <sched.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/boardctl.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <nuttx/fs/fs.h>
#include <nuttx/semaphore.h>
#include <unistd.h>
#include <arch/board/board.h>
#include <nuttx/sensors/bmi088.h>

int                   bmi088_acc_fd;
int                   bmi088_gyro_fd;
struct acc_gyro_st_s  acc_gyro;

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static int bmi088_acc_irq_handler(int irq, FAR void *context, FAR void *arg)
{
  sem_t *sem;

  sem = sem_open("/acc_semphore", O_CREAT, S_IRUSR | S_IWUSR, 1);
  if (sem == SEM_FAILED) 
    {
      return 1;
    }
  sem_post(sem);
  return 0;
}

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static void bmi088_acc_irq_handler_process(void)
{
  sem_t *sem;

  sem = sem_open("/acc_semphore", O_CREAT, S_IRUSR | S_IWUSR, 1);
  if (sem == SEM_FAILED) 
    {
      return;
    }

  while (1)
    {
      sem_wait(sem);
      read (bmi088_acc_fd , &acc_gyro, sizeof(struct acc_gyro_st_s));
    }
}

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static int bmi088_gyro_irq_handler(int irq, FAR void *context, FAR void *arg)
{
  sem_t *sem;

  sem = sem_open("/gyro_semphore", O_CREAT, S_IRUSR | S_IWUSR, 1);
  if (sem == SEM_FAILED) 
    {
      return 1;
    }
  sem_post(sem);
  return 0;
}

/****************************************************************************
 * Public Functions
 ****************************************************************************/

static void bmi088_gyro_irq_handler_process(void)
{
  sem_t *sem;

  sem = sem_open("/gyro_semphore", O_CREAT, S_IRUSR | S_IWUSR, 1);
  if (sem == SEM_FAILED) 
    {
      return;
    }

  while (1)
    {
      sem_wait(sem);
      read (bmi088_gyro_fd, &acc_gyro, sizeof(struct acc_gyro_st_s));
    }
}

/****************************************************************************
 * Name: bmi088_main
 *
 * Description:
 *   This is the main logic for the case of the BMI088 task.  It will perform
 *   one-time NSH initialization and start an interactive session on the
 *   current console device.
 *
 ****************************************************************************/

int main(int argc, FAR char *argv[])
{
  int ret = 0;

  bmi088_acc_fd  = open("dev/spi2/acc" , O_RDWR);
  bmi088_gyro_fd = open("dev/spi2/gyro", O_RDWR);

  bmi088_acc_irq_handler_register (bmi088_acc_irq_handler , &acc_gyro);
  bmi088_gyro_irq_handler_register(bmi088_gyro_irq_handler, &acc_gyro);  

  pid_t pid;

  pid = fork();

  if (pid == -1)
    {
      return 1;
    }
  else if (pid == 0)
    {
      bmi088_gyro_irq_handler_process();
    }
  else
    {
      bmi088_acc_irq_handler_process();
    }

  return ret;
}

```

nuttxpr commented:
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This PR does not meet the NuttX requirements. The provided information is insufficient and lacks crucial details. Here's a breakdown of the issues and how to address them:

**Summary:**

* **Insufficient Justification:**  "There is no driver... I am using this driver, so it is just submitted" is not a valid justification. Explain the *benefit* of adding this driver.  Who else might use it?  What functionality does it unlock?
* **Missing Technical Details:** The summary mentions SPI support and STM32G4, but doesn't connect them to the BMI088 driver itself.  Clarify how the driver interacts with the SPI bus and what specific changes were made to the STM32G4 SPI driver (if any).  Be explicit.
* **Missing Links:**  Provide links to the actual PR and any related issues.


**Impact:**

* **Vague Statements:** "Adds a new device driver function, convenient for everyone to use" is too general.  Specify which users will benefit and how.  Does this enable new applications or improve existing ones?
* **Incomplete Analysis:**  The impact section is almost entirely empty.  Address *all* the points: Impact on the user, build, hardware, documentation, security, and compatibility.  Even if the answer is "NO," state it explicitly and briefly explain why.  For example, "Impact on build: NO.  No changes to Kconfig or build system are required."


**Testing:**

* **Insufficient Host Information:**  Specify the *exact* build host OS, version, compiler, and version. Example: "Linux Ubuntu 20.04, GCC 9.4.0".
* **Insufficient Target Information:** Be more specific about the target. "STM32G474" is a microcontroller family, not a board.  Provide the board name or configuration.  Example: "STM32G474RE Nucleo board".  Also mention if you tested on a simulator (QEMU).
* **Missing "Before" Logs:** You haven't provided any "before" logs. This makes it impossible to assess the impact of the changes.
* **Unclear "After" Logs:** The provided code snippet is from the test application, not a log. Show actual output from the driver, demonstrating that it's functioning correctly (e.g., sensor readings).
* **Lack of Test Cases:** The test application seems basic.  Demonstrate testing of different driver functionalities (different sensor ranges, data rates, etc.).  Consider adding error handling tests.


**Example of a Better Summary:**

> This PR adds a new driver for the Bosch BMI088 accelerometer and gyroscope sensor, utilizing the SPI bus.  This sensor is commonly used in robotics and drone applications and provides high-precision inertial measurement. The driver supports [list specific features: e.g., various data rates, power modes, interrupt generation].  This PR also includes updates to the STM32G4 series SPI driver to support the BMI088's specific communication requirements. This change is necessary because NuttX currently lacks support for this sensor. Related Issue: [#Issue Number](Link to Issue)


**Example of a Better Impact Section:**

* **Impact on user:** YES. Users can now interface with the BMI088 sensor through the provided driver.  Example applications are provided in `apps/examples/bmi088`.
* **Impact on build:** YES.  Users will need to enable the BMI088 driver in their configuration using `CONFIG_SENSORS_BMI088`.
* **Impact on hardware:** YES.  This driver currently supports the STM32G4 series MCUs connected to the BMI088 via SPI.
* **Impact on documentation:** YES.  Documentation for the driver is included in the README file within the driver directory.  Updates to the board documentation may be necessary to describe the BMI088 connection.
* **Impact on security:** NO.  This driver does not introduce any new security vulnerabilities.
* **Impact on compatibility:** NO.  This driver is a new addition and does not affect existing functionality.



By addressing these points, you'll significantly improve the quality of your PR and increase the likelihood of it being accepted. Remember to be thorough, precise, and provide all the necessary information to allow reviewers to understand and evaluate your changes effectively.

JerrySlhao commented:
add bmi088 device used spi,now support spi. By the way, the STM32G4 series SPI driver is added to the SPI driver of STM32.

## Summary

This PR adds a new driver for the Bosch BMI088 accelerometer and gyroscope sensor, utilizing the SPI bus. This sensor is commonly used in robotics and drone applications and provides high-precision inertial measurement. The driver support users in the application layer through the ioctl to set the accelerometer and gyroscope range, data output rate, etc. This PR also includes updates to the STM32G4 series SPI driver to support the BMI088's specific communication requirements. This change is necessary because NuttX currently lacks support for this sensor. 

## Impact

### Is new feature added? 
Yes,Add a New Feature.

### Impact on user (will user need to adapt to change)?  
Yes,User need to open the macro definition.

CONFIG_SENSORS=y
CONFIG_SENSORS_BMI088=y
CONFIG_SENSORS_BMI088_APP=y
CONFIG_SENSORS_BMI088_SPI=y
CONFIG_STM32_SPI2=y
CONFIG_STM32_SPI=y
CONFIG_FS_NAMED_SEMAPHORES=y
CONFIG_FS_NAMED_SEMAPHORES_VFS_PATH="/var/sem"

### Impact on build (will build process change)?  
No.

### Impact on hardware (will arch(s) / board(s) / driver(s) change)?  
Yes.
arch/arm/src/stm32/stm32_spi.c: add macro to support stm32g4.
drivers/sensor: add bmi088.c、bmi088_base.c、bmi088_base.h、bmi088_uorb.c
include/nuttx/sensor: add bmi088.h

### Impact on documentation (is update required / provided)?
No.

### Impact on security (any sort of implications)?  
No.

### Impact on compatibility (backward/forward/interoperability)?
No.

## Testing

host machine: ubuntu20.04 
complier and version: gcc 9.4.0
board: myself board and chip type is stm32g474ret6

![image](https://github.com/user-attachments/assets/b4403bd2-4a25-4d5a-9543-ba0e35b0e997)

raiden00pl commented on include/nuttx/sensors/bmi088.h:
```diff
@@ -0,0 +1,217 @@
+
+/* IOCTL Commands ***********************************************************/
+
+#define SNIOC_ACC_GET_CHIPID              _SNIOC(0x0001)
```
what is the purpose of these? access to the system registers shouldn't be available to user-space in this way. It's not portable at all.

raiden00pl replied:
This should be done with some kind of portable interface that can be shared by all other implementations of sensors (in this case gyro/accel).
I think you should also remove the `character dev` implementation from this patch. This approach is not portable at all and we should avoid it. Otherwise, the technological debt of this project will grow indefinitely when it comes to sensors implementation. It's better to use the new sensor framework (uorb) which is standardized interface and can be also used as standard "character dev" interface".

Look for example how sensor-specific control is done for `drivers/sensors/ds18b20_uorb.c` in `ds18b20_control()`

JerrySlhao replied:
Yes, let me take a look at the uorb interface first

raiden00pl commented on include/nuttx/sensors/bmi088.h:
```diff
@@ -0,0 +1,217 @@
+#include <nuttx/config.h>
+#include <nuttx/fs/ioctl.h>
+
+#if defined(CONFIG_SENSORS_BMI088) || defined(CONFIG_SENSORS_BMI088_SCU)
```
```suggestion
```

raiden00pl commented on include/nuttx/sensors/bmi088.h:
```diff
@@ -0,0 +1,217 @@
+}
+#endif
+
+#endif /* CONFIG_SENSORS_BMI088 */
```
```suggestion
```

JerrySlhao commented on drivers/sensors/Kconfig:
```diff
@@ -1758,48 +1828,4 @@ config SENSOR_AMG88XX_ADDR
 
 endif # SENSORS_AMG88XX
 
-config SENSORS_FS3000
```
ok

JerrySlhao commented on drivers/sensors/Make.defs:
```diff
@@ -291,10 +300,6 @@ ifeq ($(CONFIG_SENSORS_LTR308),y)
   CSRCS += ltr308_uorb.c
 endif
 
-ifeq ($(CONFIG_SENSORS_FS3000),y)
```
ok

JerrySlhao commented:
Hello, I have modified this part of the code and added the uorb interface. The uorb interface has also been tested, and the application layer can successfully read the sensor data.
